### PR TITLE
[Php74][Privatization] Refactor ParentPropertyLookupGuard to use ClassReflection to reduce IO

### PR DESCRIPTION
**Pull request by samsonasik**
Ref https://github.com/rectorphp/rector-src/pull/3992#discussion_r1208005540

### CODE REVIEW

File: rules/Php74/Guard/PropertyTypeChangeGuard.php [hunk 1]
```diff
@@ -76,15 +76,10 @@ private function isSafeProtectedProperty(ClassReflection $classReflection, Prope
             return false;
         }
 
-        $class = $this->classLikeAstResolver->resolveClassFromClassReflection($classReflection);
-        if (! $class instanceof Class_) {
+        if (! $classReflection->isFinalByKeyword()) {
```
Comment (samsonasik) [inline rules/Php74/Guard/PropertyTypeChangeGuard.php:76]: `isFinalByKeyword()` here is needed to ensure read final from class, not docblock, same with `$class->isFinal()` usage which read from real `final` class.

File: rules/Privatization/Guard/ParentPropertyLookupGuard.php [hunk 2]
```diff
@@ -51,8 +45,9 @@ public function isLegal(Property $property, Class_ $class): bool
             return false;
         }
 
-        if (! $class->extends instanceof Name) {
-            return true;
+        $nativeReflection = $classReflection->getNativeReflection();
+        if (! $nativeReflection->getParentClass()) {
```
Comment (samsonasik) [inline rules/Privatization/Guard/ParentPropertyLookupGuard.php]: extends class may be not autoloaded, so need native reflection to ensure it has parent class.

Comment (samsonasik) [reply rules/Privatization/Guard/ParentPropertyLookupGuard.php]: It seems the NativeReflection is BetterReflection class, so it got test error on not autoload part:

```diff
 final class KeepParentNotAutoloaded extends NotAutoloadedParent
 {
-    protected $value = 100;
+    private $value = 100;
 }
```

Comment (samsonasik) [reply rules/Privatization/Guard/ParentPropertyLookupGuard.php]: Fixed by pull from `BetterReflectionClass` object https://github.com/rectorphp/rector-src/pull/3993/commits/36f291f0d490cad21d6d4e32b3e5aa5f411ac23c

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I am merging it ;)

**Comment by TomasVotruba** [comment]
Thanks a lot :+1:

**Pull request merged**